chained/selector_test.go

```diff
@@ -0,0 +1,134 @@
+package chained
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSelectorDial(t *testing.T) {
+	tests := []struct {
+		name     string
+		assertFn func(t *testing.T)
+	}{
+		{name: "dial success", assertFn: assertSelectorSuccess},
+		{name: "dialer advance", assertFn: assertSelectorAdvance},
+		{name: "fail to dial", assertFn: assertSelectorFailToDial},
+		{name: "unsupported addr", assertFn: assertSelectorUnsupportedAddr},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.assertFn(t)
+		})
+	}
+}
+
+func assertSelectorSuccess(t *testing.T) {
+	selector := NewSelector(
+		[]Dialer{
+			&mockDialer{name: "dialer-0"},
+			&mockDialer{name: "dialer-1"},
+		},
+	)
+	_, err := selector.Dial("tcp", selectorTestAddr)
+	assert.NoError(t, err)
+	assert.Equal(t, int64(0), selector.active.Load(), "dialer 0 should still be active dialer")
+}
+
+func assertSelectorAdvance(t *testing.T) {
+	selector := NewSelector(
+		[]Dialer{
+			&mockDialer{name: "failingDialer", forceFail: true},
+			&mockDialer{name: "succeedingDialer"},
+		},
+	)
+	_, err := selector.Dial("tcp", selectorTestAddr)
+	assert.NoError(t, err)
+	assert.Equal(t, int64(1), selector.active.Load(), "selector should have advanced to dialer 1")
+}
+
+func assertSelectorFailToDial(t *testing.T) {
+	selector := NewSelector(
+		[]Dialer{
+			&mockDialer{name: "failingDialer-0", forceFail: true},
+			&mockDialer{name: "failingDialer-1", forceFail: true},
+		},
+	)
+	_, err := selector.Dial("tcp", selectorTestAddr)
+	assert.Error(t, err)
+}
+
+func assertSelectorUnsupportedAddr(t *testing.T) {
+	// we need to test that Dial fails when no dialer supports the addr to avoid infinite loop
+	selector := NewSelector(
+		[]Dialer{
+			&mockDialer{name: "Microsoft"},
+			&mockDialer{name: "Microsoft still"},
+		},
+	)
+	_, err := selector.Dial("", "working windows updates")
+	assert.Error(t, err)
+}
+
+const selectorTestAddr = "http://test.addr"
+
+type mockDialer struct {
```

Comment:
This suggestion is purely optional and I'm just adding this solely for discussion purposes. Can we use https://github.com/uber-go/mock for generating interfaces mocks? It's generally easier to maintain and generate them whenever we need to update them (using `go generate ./...`). But again, this can be ignored, the current solution woks and it's good.

Reply:
I'm not really familiar with it so I'll have to check it out. But, if it makes creating mocks easier, then I'm all for it.